pkg/lib/serverTime.js

```diff
@@ -321,13 +302,109 @@
+            data.split("\n").forEach(function(line) {
+                const parts = line.split(" ");
+                if (parts[0] == "server")
+                    servers.push(parts[1]);
+            });
+            return servers;
+        }
+
+        return enabled_file.read()
+                .then(data => {
+                    if (data) {
+                        result.servers = parse_servers(data);
+                        result.enabled = true;
+                        return Promise.resolve(result);
+                    } else {
+                        return disabled_file.read()
+                                .then(data => {
+                                    result.servers = parse_servers(data);
+                                    result.enabled = false;
+                                    return Promise.resolve(result);
+                                });
+                    }
+                });
+    }
+
+    function set_custom_ntp_chronyd(servers, enabled) {
+        const enabled_file = cockpit.file("/etc/chrony-cockpit-sources/custom.sources", { superuser: "try" });
+        const disabled_file = cockpit.file("/etc/chrony-cockpit-sources/custom.disabled", { superuser: "try" });
+
+        const text = "# This file is automatically generated by Cockpit\n\n" + servers.map(s => `server ${s}\n`).join("");
+
+        function ensure_sourcedir() {
+            function add_sourcedir(data) {
+                if (data.indexOf("sourcedir /etc/chrony-cockpit-sources") == -1)
+                    data += "\n# Sources added by the Cockpit web console\nsourcedir /etc/chrony-cockpit-sources\n";
+                return data;
+            }
+            return cockpit.file("/etc/chrony.conf", { superuser: "try" }).modify(add_sourcedir);
+        }
+
+        return cockpit.spawn(["mkdir", "-p", "/etc/chrony-cockpit-sources"], { superuser: "try" })
+                .then(() => {
+                    if (enabled)
+                        enabled_file.replace(text).then(() => disabled_file.replace(null)).then(ensure_sourcedir);
+                    else
+                        disabled_file.replace(text).then(() => enabled_file.replace(null));
+                });
+    }
+
+    self.get_custom_ntp = function () {
+        /* We support editing the configuration of systemd-timesyncd
+         * or chronyd, by dropping a config file snippet that we
+         * assume full control over.
+         *
+         * XXX - figure out a good way to detect which one to
+         *       configure. Look into ntp-units.d?
+         */
+
+        if (!timedate1_service.exists || timedate1_service.unit.Id !== "systemd-timedated.service") {
+            console.log("systemd-timedated not in use, ntp server configuration not supported");
+            return Promise.resolve({ supported: false, servers: [], enabled: false });
+        }
+
+        if (timesyncd_service.enabled)
+            return get_custom_ntp_timesyncd();
```

Comment:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17992-20221206-194101-a7fca0c4-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)